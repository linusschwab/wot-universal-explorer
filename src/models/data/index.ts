export {Data} from './Data'
export {InputData} from './InputData'
export {OutputData} from './OutputData'
