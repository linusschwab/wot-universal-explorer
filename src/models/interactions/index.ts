export {InteractionData} from './InteractionData';
export {InteractionPattern} from './InteractionPattern';
export {Action} from './Action';
export {Event} from './Event';
export {Property} from './Property';
export {ISubscriber} from './ISubscriber';
