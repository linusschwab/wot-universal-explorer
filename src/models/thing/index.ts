export {Thing} from './Thing';
export {ThingsManager} from './ThingsManager';
