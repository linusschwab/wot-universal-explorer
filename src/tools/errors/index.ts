export {InteractionError} from "./InteractionError";
export {TimeoutError} from "./TimeoutError";
