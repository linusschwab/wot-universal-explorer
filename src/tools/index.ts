export {MozillaTDParser} from "./MozillaTDParser";
export {OpenAPIEncoder} from "./OpenAPIEncoder";
export {TDEncoder} from "./TDEncoder";
export {TDParser} from "./TDParser";
