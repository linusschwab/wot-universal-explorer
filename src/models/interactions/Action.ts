import {Interaction} from "./Interaction";
import {InputData, OutputData} from "../data";


export class Action extends Interaction {

    public inputData: InputData;

    constructor(name: string, outputData: OutputData, inputData: InputData) {
        super(name);
        this.outputData = outputData;
        this.inputData = inputData;
    }

    public invoke() {
        // TODO: Implement
        return true;
    }

    public toString() {
        return this.name + ' action';
    }

}
