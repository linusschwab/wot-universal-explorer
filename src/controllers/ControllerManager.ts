import {IndexController, TDController, ThingsController} from "./index";
import {ThingsManager} from "../models/thing";


export class ControllerManager {

    public td: TDController;
    public things: ThingsController;
    public index: IndexController;

    constructor(things: ThingsManager) {
        this.td = new TDController(things);
        this.things = new ThingsController(things);
        this.index = new IndexController();
    }
}
