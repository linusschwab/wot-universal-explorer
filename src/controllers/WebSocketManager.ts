import * as WebSocket from "ws";
import * as http from "http";
import {Thing, ThingsManager} from "../models/thing";
import {ControllerManager} from "../controllers";
import {RequestError, ThingError} from "../tools/errors";


export class WebSocketManager {

    public server: WebSocket.Server;
    private controllers: ControllerManager;
    private things: ThingsManager;

    public static port = 8080;

    constructor(controllers: ControllerManager, things: ThingsManager) {
        this.controllers = controllers;
        this.things = things;

        this.server = new WebSocket.Server({ port: WebSocketManager.port });
        this.server.on('connection', (ws, req) => this.connection(ws, req));
    }

    public async connection(ws: WebSocket, req: http.IncomingMessage) {
        try {
            const thing = this.getThing(req);

            ws.on('message', async (data) => this.handleMessage(ws, data, thing));

            ws.on('error', () => {
                // TODO: Remove subscriber
            });
            ws.on('close', () => {
                // TODO: Remove subscriber
            });

            ws.send('connected to ' + thing.name);
        } catch (e) {
            this.handleError(ws, e);
        }
    }

    private handleMessage(ws: WebSocket, data: WebSocket.Data, thing: Thing) {
        try {
            const message = this.parseMessage(data);
            this.controllers.things.wsRoutes(ws, message, thing);
        } catch (e) {
            this.handleError(ws, e);
        }
    }

    public parseMessage(data: WebSocket.Data) {
        let message: any;

        try {
            message = JSON.parse(<string>data);
        } catch (e) {
            throw new RequestError('Unable to parse message');
        }

        if (!message.hasOwnProperty('messageType') || !message.hasOwnProperty('data')) {
            throw new RequestError('Invalid message format');
        }

        return message;
    }

    public getThing(req: http.IncomingMessage) {
        // Match first part of path
        let path = req.url.match(/\/(.*?)(?:\/|$)/g);

        if (path && path.length == 1) {
            let name = path[0];
            return this.things.getThing(name);
        }
        throw new ThingError('Invalid thing name');
    }

    public handleError(ws: WebSocket, e: Error) {
        if (e instanceof ThingError) {
            ws.send(JSON.stringify({
                messageType: 'error',
                data: {
                    status: '404',
                    message: e.message
                }
            }));
            ws.close();
        } else if (e instanceof RequestError) {
            ws.send(JSON.stringify({
                messageType: 'error',
                data: {
                    status: '400',
                    message: e.message
                }
            }));
        } else {
            throw e;
        }
    }
}
