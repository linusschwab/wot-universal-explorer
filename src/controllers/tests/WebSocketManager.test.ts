import * as http from "http";
import * as WebSocket from "ws";

import {WebSocketManager} from "../WebSocketManager";
import {Thing, ThingsManager} from "../../models/thing";
import {ControllerManager} from "../ControllerManager";
import {InteractionError, RequestError, ThingError, TimeoutError} from "../../tools/errors";
import {Action, Event, InteractionData, Property} from "../../models/interactions";


// Create mocks
jest.mock('ws');

let mockWs: WebSocket;


let things = new ThingsManager();
let controllers = new ControllerManager(things);
const ws = new WebSocketManager(null, controllers, things);


beforeAll(() => {
    const MockWebSocket = jest.fn<WebSocket>(() => ({
        send: jest.fn(),
        close: jest.fn()
    }));
    mockWs = new MockWebSocket();
});

beforeEach(() => {
    jest.clearAllMocks();
});

describe('parse message', () => {
    test('empty message throws error', async () => {
        const data = JSON.stringify("");

        expect(() => {
            WebSocketManager.parseMessage(data)
        }).toThrowError(RequestError);
    });

    test('invalid message throws error', async () => {
        const data = JSON.stringify({
                messageType: "data missing"
        });

        expect(() => {
            WebSocketManager.parseMessage(data)
        }).toThrowError(RequestError);
    });

    test('valid message returns javascript object', async () => {
        const data = {
            messageType: "testMessage",
            data: {
                test: "value"
            }
        };
        const message = JSON.stringify(data);

        expect(WebSocketManager.parseMessage(message)).toEqual(data);
    });
});

describe('get thing', () => {
    const thing = new Thing('Test Thing', 'thing');
    things.addThing(thing);

    test('invalid url throws error', async () => {
        const req = new http.IncomingMessage(null);
        req.url = '/test';

        expect(() => {
            ws.getThing(req)
        }).toThrowError(ThingError);
    });

    test('valid url returns correct thing', async () => {
        const req = new http.IncomingMessage(null);
        req.url = '/test-thing';

        expect(ws.getThing(req)).toBe(thing);
    });
});

describe('subscriber notifications', () => {
    const data = new InteractionData('Test Data');

    test('property notification sends correct message', async () => {
        const property = new Property('Test Property', null, false, true);

        WebSocketManager.notify(mockWs, property, data);

        expect(mockWs.send).toMatchSnapshot();
    });

    test('action notification sends correct message', async () => {
        const action = new Action('Test Action', null, null);

        WebSocketManager.notify(mockWs, action, data);

        expect(mockWs.send).toMatchSnapshot();
    });

    test('event notification sends correct message', async () => {
        const event = new Event('Test Event', null);

        WebSocketManager.notify(mockWs, event, data);

        expect(mockWs.send).toMatchSnapshot();
    });
});

describe('error handling', () => {
    test('thing error sends error message and closes connection', async () => {
        const error = new ThingError('Error message');

        WebSocketManager.handleError(mockWs, error);

        expect(mockWs.send).toMatchSnapshot();
        expect(mockWs.close).toHaveBeenCalledTimes(1);
    });

    test('timeout error sends error message', async () => {
        const error = new TimeoutError('Error message');

        WebSocketManager.handleError(mockWs, error);

        expect(mockWs.send).toMatchSnapshot();
    });

    test('request error rejects message', async () => {
        const error = new RequestError('Error message');

        WebSocketManager.handleError(mockWs, error);

        expect(mockWs.send).toMatchSnapshot();
    });

    test('interaction error rejects message', async () => {
        const error = new InteractionError('Error message');

        WebSocketManager.handleError(mockWs, error);

        expect(mockWs.send).toMatchSnapshot();
    });
});
