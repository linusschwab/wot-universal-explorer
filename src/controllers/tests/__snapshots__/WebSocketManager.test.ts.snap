// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error handling interaction error rejects message 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"messageType\\":\\"error\\",\\"data\\":{\\"status\\":\\"400\\",\\"message\\":\\"Error message\\"}}",
    ],
  ],
}
`;

exports[`error handling request error rejects message 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"messageType\\":\\"error\\",\\"data\\":{\\"status\\":\\"400\\",\\"message\\":\\"Error message\\"}}",
    ],
  ],
}
`;

exports[`error handling thing error sends error message and closes connection 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"messageType\\":\\"error\\",\\"data\\":{\\"status\\":\\"404\\",\\"message\\":\\"Error message\\"}}",
    ],
  ],
}
`;

exports[`error handling timeout error sends error message 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"messageType\\":\\"error\\",\\"data\\":{\\"status\\":\\"408\\",\\"message\\":\\"Error message\\"}}",
    ],
  ],
}
`;

exports[`subscriber notifications action notification sends correct message 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"messageType\\":\\"action\\",\\"data\\":{\\"Test Action\\":\\"Test Data\\"}}",
    ],
  ],
}
`;

exports[`subscriber notifications event notification sends correct message 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"messageType\\":\\"event\\",\\"data\\":{\\"Test Event\\":\\"Test Data\\"}}",
    ],
  ],
}
`;

exports[`subscriber notifications property notification sends correct message 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"messageType\\":\\"propertyStatus\\",\\"data\\":{\\"Test Property\\":\\"Test Data\\"}}",
    ],
  ],
}
`;
